### PR TITLE
Add sequence weighting

### PR DESCRIPTION
**Pull request by priyappillai**
ADAPT currently bases how much coverage it has of a virus on the percent of genomic sequences in the database that are covered. While this works well if the sequences represent a random sample of the population, this is not always the case due to sampling biases.

For the case when ADAPT is designing an assay across a taxon with multiple subtaxa, each with different levels of sampling, ADAPT can create assays that only cover a highly overrepresented subtaxon and no other subtaxa. While having more sequences in the database is representative of a subtaxon's relative importance, it should not cause other subtaxa to be treated as unimportant.

This PR introduces weighting sequences, both automatically and manually, to help account for this problem. Manual weights can be input with `--weight-sequences`, and weights can be determined automatically with `--weight-by-log-size-of-subtaxa`. Automatic weighting uses the log of the number of sequences within a subtaxa as a heuristic of how important that group is.

`adapt/utils/weight.py`
- Create file
- Add functions
  - `normalize`
  - `percentile` (weighted)
  - `weight_by_log_group`

`adapt/utils/tests/test_weight.py` 
- Create file

`adapt/utils/seq_io.py`
- Add `read_sequence_weights`

`adapt/alignment.py`
- Add normalized weights as a property of Alignment objects
- Add `seq_idxs_weighted` function to get the sum of the weights of a subset of sequences
- In `construct_guide`, change `num_needed` to `percent_needed`
- In `determine_consensus_sequence`/`position_entropy`/`base_percentages`, use weights to calculate base percentages rather than counts
- In determine_most_common_sequences, use weights for sequence percentages rather than counts

`adapt/tests/test_alignment.py`
- Add tests for `seq_idx_weighted`
- Add tests for weighted `construct_guide`/`determine_consensus_sequence`/`position_entropy`/`base_percentages`/`determine_most_common_sequences`

`adapt/prepare/align.py`
- Move reference accession check to before the loop in `curate_against_ref`

`adapt/prepare/prepare_alignment.py`
- Change prepare alignments to use 0 for the 'tax ID' and the unaligned FASTA file name for the 'segment'
- Output weights

`adapt/prepare/ncbi_neighbors.py`
- Add functions:
  - `ncbi_detail_taxonomy_url`
  - `ncbi_search_taxonomy_url`
  - `parse_taxonomy_xml_for_taxid`
  - `parse_taxonomy_xml_for_rank`
  - `fetch_taxonomies`
  - `get_taxid`
  - `get_rank`
  - `get_subtaxa_groups`

`adapt/prepare/tests/test_ncbi_neighbors.py`
- Test NCBI taxonomy url generation
- Test fetching taxonomies, getting rank, and getting subtaxa groups

`adapt/guide_search.py`
- In guide_set_activities_percentile, use weighted percentile
- In `guide_set_activities_expected_value`/`guide_set_activities_expected_value_per_guide`/`guide_activities_expected_value`/`obj_value`/`_ground_set_with_activities_memoized`/`_analyze_guides`/`_analyze_guides_memoized`/`add_guide, use weighted average
- In `_construct_guide_memoized`/`_find_optimal_guide`/`_find_guides_in_window`/`add_guide_to_cover`/`score_collection_of_guides`, use percent_needed rather than num_needed
- In `total_frac_bound_by_guides`, use weighted total (using `seq_idxs_weighted`)

`adapt/tests/test_guide_search.py`
- Modify tests to use percent needed rather than num needed
- Use almost equal for percent needed for floating points issues

`bin/design.py`
- Change prepare alignments to use 0 for the 'tax ID' and the unaligned FASTA file name for the 'segment'
- Set up preparation steps to also include outputting weights
- Add argument `--weight-by-log-size-of-subtaxa` to be passed to `prepare_alignment.prepare_for`

`bin/tests/test_design.py`
- Add weighted argument type
- Test weighted FASTA input 
- Test auto weighted sequences

### CODE REVIEW

File: adapt/utils/seq_io.py [hunk 1]
```diff
@@ -441,6 +441,37 @@ def read_taxonomy_specificity_ignore(fn):
...
+        1) a name of the sequence in the input FASTA
+        2) a weight as a floating point number
+
+    Args:
+        fn: path to TSV file, where each row corresponds to sequence
+
+    Returns:
+        dictionary of {sequence name: unnormalized weight}
+    """
+    weights = defaultdict(lambda: 1)
```
Comment (haydenm) [inline adapt/utils/seq_io.py:457]: Not sure, but I think a downside of returning a `defaultdict()` might be that if a sequence is not given a weight in the input, it will automatically be considered to have (unnormalized) weight of 1?

File: bin/design.py [hunk 2]
```diff
@@ -1333,6 +1357,15 @@ def __call__(self, parser, namespace, values, option_string=None):
...
               "i.e., year n is given B*(desired cover fraction of year n+1)."))
+    input_fasta_subparser.add_argument('--weight-sequences', nargs='+',
+        help=("If set, file path to TSV giving a weight to each input "
+              "sequence (col 1 is sequence name matching that in the "
+              "respective input FASTA, col 2 is weight). If more than one "
+              "input FASTA is given, the same number of output TSVs must be "
+              "given; each output TSV corresponds to an input FASTA. The "
+              "weights will be normalized and used when calculating expected "
+              "activity and percent coverage. Any sequence not listed will be "
+              "given a default weight of 1."))
```
Comment (haydenm) [inline bin/design.py:1368]: This addresses my question in an earlier comment in `seq_io.py`. Why is it the right option, since 1 can mean anything based on the magnitudes of the weights?

File: bin/design.py [hunk 3]
```diff
@@ -813,28 +830,26 @@ def run(args):
 
     logger.info("Running design.py with arguments: %s", args)
 
-    # Set NCBI API key
```
Comment (haydenm) [inline bin/design.py:816]: Why the deletion?

Comment (haydenm) [reply bin/design.py:816]: Nevermind, I see now that I think it's because you merged two blocks of code.

File: bin/design.py [hunk 4]
```diff
@@ -1372,6 +1405,13 @@ def __call__(self, parser, namespace, values, option_string=None):
               "for guides (and PRIMER_COVER_FRAC for primers). Each preceding "
               "year receives a desired cover fraction that decays by B -- "
               "i.e., year n is given B*(desired cover fraction of year n+1)."))
+    input_auto_common_subparser.add_argument('--weight-by-log-size-of-subtaxa',
```
Comment (haydenm) [inline bin/design.py:1408]: May want to check somewhere that this and `--weight-sequences` are not both set (it seems, if they are, that the log size of subtext will override the file?).

File: adapt/utils/weight.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+import math
+import numpy as np
+
+from adapt.prepare import ncbi_neighbors
+
+__author__ = 'Priya P. Pillai <ppillai@broadinstitute.org>'
+
+
+def weight_by_log_group(groups):
+    """Weight sequences by the log of the number of sequences in the group
```
Comment (haydenm) [inline adapt/utils/weight.py:14]: This header should be more descriptive, i.e., every sequence falls into exactly one group (maybe check/assert this?), a group's weight is proportional to the log of its size, and every sequence in the group gets that group's weight divided by the size of the group.

File: adapt/utils/weight.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    # ordering
+    q_pos = {q_i/100: pos for pos, q_i in enumerate(q)}
+    # Order the requested percentiles to iterate through them
+    ordered_qs = sorted(q_pos.keys())
+    curr_q = ordered_qs.pop(0)
+    # Keep track of the index of the item in the original percentile list that
+    # matches the requested percentile (in the same order as the original qs)
+    p_idxs = [-1 for _ in q]
+    while len(ordered_qs) > 0 and i < len(sorted_activities_idxs):
+        # Since sequence weights are normalized, the sum of weights of
```
Comment (haydenm) [inline adapt/utils/weight.py:87]: Check that they are indeed normalized and/or include that in the description of `seq_norm_weights`.

File: adapt/utils/weight.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    if normalization_factor <= 0:
+        raise ValueError("Weights must have a sum greater than or equal to 0; "
+            "the weights currently sum to %f" %(normalization_factor))
+
+    return {seq_name: (sequence_weights[seq_name] / normalization_factor)
+            for seq_name in seq_names}
+
+
+def percentile(activities, q, seq_norm_weights):
+    """Take a weighted percentile of activities, using 'lower' interpolation
```
Comment (haydenm) [inline adapt/utils/weight.py:60]: I think this should either explain a bit more what we mean by weighted percentile or link to something that does, e.g., [https://en.wikipedia.org/wiki/Percentile#The_weighted_percentile_method](https://en.wikipedia.org/wiki/Percentile#The_weighted_percentile_method)

File: adapt/utils/tests/test_weight.py [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+                                                   'group_2': ['B', 'C', 'D']})
+        test_norm_weights = weight.normalize(test_weights,
+                                             ['A', 'B', 'C', 'D'])
+        self.assertAlmostEqual(test_norm_weights['A'], 1/3)
+        self.assertAlmostEqual(test_norm_weights['B'], 2/9)
+        self.assertAlmostEqual(test_norm_weights['C'], 2/9)
+        self.assertAlmostEqual(test_norm_weights['D'], 2/9)
+
+
+    def test_percentile(self):
```
Comment (haydenm) [inline adapt/utils/tests/test_weight.py:67]: This is such a complicated function! I'm going to ask for many more test examples :)  Both standard uses and edge cases.

**Review by haydenm — CHANGES_REQUESTED** [review]
This is super impressive work @priyappillai, in everything from the tricky software implementations (e.g., NCBI parsing) through more complicated functions (e.g., weighted percentiles). And it will definitely play, I think, a key role in the short/medium term in how we go about correcting for sampling biases during design.

File: adapt/guide_search.py [hunk 9]
```diff
@@ -477,8 +477,9 @@ def guide_set_activities_expected_value_per_guide(self, window_start,
             # best_guide serves as a mask to only take the mean of the
             # sequences for which this guide has the maximum activity
             best_guide = per_gd_activities[gd_seq] == activities
-            per_gd_expected_activities[gd_seq] = np.mean(
-                per_gd_activities[gd_seq][best_guide])
+            per_gd_expected_activities[gd_seq] = np.average(
+                per_gd_activities[gd_seq][best_guide],
+                weights=np.array(self.aln.seq_norm_weights)[best_guide])
```
[the remark this answers is not in the document]
Comment (priyappillai) [reply adapt/guide_search.py:482]: It's switched to a numpy array here in order to use `best_guide` as a mask; L399 doesn't need a mask

File: adapt/alignment.py [hunk 10]
```diff
@@ -100,16 +106,29 @@ def extract_range(self, pos_start, pos_end):
         Returns:
             object of type Alignment including only the specified range
         """
-        return Alignment(self.seqs[pos_start:pos_end])
+        return Alignment(self.seqs[pos_start:pos_end],
+                         seq_norm_weights=self.seq_norm_weights)
 
     def _compute_frac_missing(self):
         """Compute fraction of sequences with missing data at each position.
```
[the remark this answers is not in the document]
Comment (priyappillai) [reply adapt/alignment.py:113]: I'm not sure entirely what you mean? It's just making sure that the weights of the original alignment are still reflected in the alignment of the smaller range. With the exception of `pick_test_targets` (which will use an alignment without weights anyway), it is exclusively used for getting alignments for guides, which should use the weights.

File: adapt/prepare/prepare_alignment.py [hunk 11]
```diff
@@ -315,6 +332,13 @@ def prepare_for(taxid, segment, ref_accs, out,
         else:
             annotations.append([])
 
+        if subtaxa_weight:
+            subtaxa_groups = ncbi_neighbors.get_subtaxa_groups(
+                seqs_aligned.keys(), subtaxa_weight)
+            sequence_weights.append(weight.weight_by_log_group(subtaxa_groups))
+        else:
+            sequence_weights.append(defaultdict(lambda: 1))
```
[the remark this answers is not in the document]
Comment (priyappillai) [reply adapt/prepare/prepare_alignment.py:340]: This is unnormalized even with `weight_by_log_group`; it will eventually be normalized when put in the alignment.

File: bin/design.py (hunk 2 above)
Comment (priyappillai) [reply bin/design.py:1368]: I think the other options would be to have it fail if there isn't a weight listed for every sequence or to set the weight of that sequence to 0. 
I don't like having it fail mostly because I can imagine that being fairly annoying if you only want to weigh a subset of sequences.
For 0 as a default, I was avoiding allowing any weights to be 0 (since you could just remove that sequence from the FASTA in that case), but I can imagine scenarios where it's useful (such as making designs for various subsets of a FASTA).
I originally picked 1 as the default because I figured it would be easiest conceptually to scale other sequences against a default of 1. 
I'm not certain whether 0 or 1 makes more sense as a default; what do you think?

Comment (haydenm) [reply bin/design.py:1368]: I was thinking to fail if a sequence is missing from the input file, but your comment about wanting to only weigh a subset of sequences makes sense and is totally reasonable. Definitely don't think 0 is the way to go. Maybe as a compromise stick with 1, but have a log message (at `WARNING` level) if ≥1 sequence is missing from the file, just to remind the user that they will be given a weight of 1?

File: adapt/prepare/prepare_alignment.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (priyappillai) [reply adapt/prepare/prepare_alignment.py:340]: Yup, L131 in design.py to be precise!

File: adapt/alignment.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (priyappillai) [reply adapt/alignment.py:113]: Looking at what this is doing, I don't actually think it should be weighted at all (especially given how it's memoized).  I might switch this back to being unweighted?

File: adapt/prepare/ncbi_neighbors.py [hunk 12]
```diff
@@ -935,26 +1025,161 @@ def fetch_metadata(accessions):
...
+    Returns:
+        numerical taxonomic ID
+    """
+    search_url = ncbi_search_taxonomy_url(taxon_name)
+    raw_search_xml = urlopen_with_tries(search_url)
+
+    return parse_taxonomy_xml_for_taxid(raw_search_xml)[0]
+
+
+def get_rank(taxid, rank):
```
[the remark this answers is not in the document]
Comment (priyappillai) [reply adapt/prepare/ncbi_neighbors.py:1127]: Yeah, I overlapped the naming of terms here. Can rename to `get_taxonomy_name_of_rank`?

File: adapt/utils/seq_io.py (hunk 1 above)
Comment (priyappillai) [reply adapt/utils/seq_io.py:457]: (Discussed in the subsequent comment)

File: bin/design.py (hunk 4 above)
Comment (priyappillai) [reply bin/design.py:1408]: `--weight-sequences` is only in the FASTA subparser, so they shouldn't be allowed to both be set

File: adapt/utils/weight.py (hunk 6 above)
Comment (priyappillai) [reply adapt/utils/weight.py:114]: Ended up checking if they were normalized, and normalizing them if they were not!

File: adapt/alignment.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (haydenm) [reply adapt/alignment.py:116]: Yeah, if it's not used for anything where weighted is needed, then unweighted makes more sense.

**Review by haydenm — APPROVED** [review]
👍  Changes look good, and I just made some minor updates to the README.
